src/backend/joanie/badges/providers/obf.py

```diff
@@ -0,0 +1,468 @@
+"""OpenBadgeFactory provider."""
+
+import json
+import logging
+import re
+from collections.abc import Iterable
+from functools import cached_property
+from typing import Literal, Optional
+from urllib.parse import urljoin
+
+# pylint: disable=no-name-in-module
+import requests
+from pydantic import (
+    BaseModel,
+    EmailStr,
+    HttpUrl,
+    NoneStr,
+    ValidationError,
+    root_validator,
+)
+
+from ..exceptions import AuthenticationError, BadgeProviderError
+from .base import BaseProvider
+
+logger = logging.getLogger(__name__)
+
+
+class OAuth2AccessToken(requests.auth.AuthBase):
+    """Add OAuth2 access token to HTTP API requests header."""
+
+    def __init__(self, access_token):
+        """Instantiate requests Auth object with generated access_token."""
+
+        self.access_token = access_token
+
+    def __call__(self, request):
+        """Modify and return the request."""
+
+        request.headers.update(
+            {
+                "Authorization": f"Bearer {self.access_token}",
+            }
+        )
+
+        return request
+
+
+class OBFAPIClient(requests.Session):
+    """Open Badge Factory API Client."""
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        *args,
+        raise_for_status: bool = False,
+        **kwargs,
+    ):
+        """Override default requests.Session instantiation to handle authentication."""
+
+        super().__init__(*args, **kwargs)
+
+        self.api_root_url = "https://openbadgefactory.com"
+        self.api_version_prefix = "v1"
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.raise_for_status = raise_for_status
+        self.auth = self._get_auth
+        self.headers.update(
+            {
+                "Content-Type": "application/json",
+            }
+        )
+
+    @cached_property
+    def _access_token(self):
+        """Request OAuth2 access token from the API backend.
+
+        We cache this property to avoid regenerating an access token at each
+        API request. This access token has a limited validity (e.g. 10h) so we
+        try to regenerate it when the API response code is 403 (see the
+        `request` overridden method).
+
+        """
+
+        url = f"{self.api_version_prefix}/client/oauth2/token"
+        response = requests.post(
+            urljoin(self.api_root_url, url),
+            json={
+                "grant_type": "client_credentials",
+                "client_id": self.client_id,
+                "client_secret": self.client_secret,
+            },
+        )
+
+        if "access_token" not in response.json():
```

Comment:
I would say there are cases when response.json() will fail so you need to first test if the request was successful?